torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -0,0 +1,223 @@
+    for (auto expr : fl->body().exprs()) {
+      is_last_op_sync_ = false;
+      if (expr->getExprType().value() == ExprType::Sync) {
+        initial_sync_ = true;
+        final_.clear();
+      } else if (expr->getExprType().value() == ExprType::ForLoop) {
+        // Recursively handle nested for-loop
+        LocalSyncInserter child_sync_inserter;
+        child_sync_inserter.handle(expr);
+        const auto& child_inputs = child_sync_inserter.all_smem_inputs();
+        const auto& child_outputs = child_sync_inserter.all_smem_outputs();
+
+        // Default - Track all smem inputs / outputs
+        all_smem_inputs_.insert(child_inputs.begin(), child_inputs.end());
+        all_smem_outputs_.insert(child_outputs.begin(), child_outputs.end());
+
+        if (!initial_sync_) {
+          // Parent - None
+          if (!child_sync_inserter.initial_sync_) {
+            // Child - None
+            // Append All Child Outputs to Parent Initial
+            initial_.insert(child_outputs.begin(), child_outputs.end());
+          } else if (child_sync_inserter.has_war_hazard_sync_) {
+            // Child - WAR race
+            // Parent first sync
+            // Inherit Child Initial / Clear Parent Final
+            initial_sync_ = true;
+            is_last_op_sync_ = true;
+            initial_.insert(
+                child_sync_inserter.initial().begin(),
+                child_sync_inserter.initial().end());
+            final_.clear();
+          } else {
+            // Child - 1+
+            // Parent first sync
+            // Inherit Child Initial + Final
+            initial_sync_ = true;
+            initial_.insert(
+                child_sync_inserter.initial().begin(),
+                child_sync_inserter.initial().end());
+            final_.insert(
+                child_sync_inserter.final().begin(),
+                child_sync_inserter.final().end());
+          }
+        } else {
+          // Parent - 1+
+          if (!child_sync_inserter.initial_sync_) {
+            // Child - None
+            // Append All Child to Parent Last
+            final_.insert(child_inputs.begin(), child_inputs.end());
+          } else if (child_sync_inserter.has_war_hazard_sync_) {
+            // Child - WAR race
+            // Clear Parent Last / Discard Child Initial
+            is_last_op_sync_ = true;
+            final_.clear();
+          } else {
+            // Child - 1+
+            // Inherit Child Final / Discard Child Initial
+            final_.insert(
+                child_sync_inserter.final().begin(),
+                child_sync_inserter.final().end());
+          }
+        }
+      } else {
+        handle(expr);
+      }
+    }
+
+    // This level of the nested for-loop may not exist in the kernel.
+    // However, subsequent levels can exist, so we handle the body of the
+    // for-loop first.
+    if (!fl->iter_domain()->isThread() && !fl->iter_domain()->isBroadcast()) {
```

Comment:
This is an example of why doing this very late in the lowering process might be more natural.

Reply:
We perform this check in [codegen.cpp #L534-L538](https://github.com/csarofeen/pytorch/blob/20_8_18_devel/torch/csrc/jit/codegen/cuda/codegen.cpp#L534-L538) as we're generating the kernel.

Do you want another stage in the lowering process that handles this check before the codegen process?

Reply:
Right - that's a problem as well, which is why I don't think this PR is the right place to change the approach (the codegen should not have that check either, it should be done part of the lowering - if you want to add a comment there go ahead)